zkevm-circuits/src/evm_circuit/op_execution/arithmetic.rs

```diff
@@ -0,0 +1,577 @@
+                        + Expression::Constant(F::from_u64(3))),
+            ];
+
+            let AddSuccessAllocation {
+                selector,
+                swap,
+                a,
+                b,
+                c,
+                carry,
+            } = &self.success;
+
+            // swap a and c if it's SUB
+            let no_swap = one - swap.expr();
+            let swap_constraints = vec![
+                swap.expr() * no_swap.clone(),
+                swap.expr() * (opcode.expr() - sub),
+                no_swap.clone() * (opcode.expr() - add),
+            ];
+
+            // add constraints
+            let add_constraints = {
+                let mut constraints = Vec::with_capacity(32);
+                // 256 * carry_out + c
+                let lhs =
+                    carry[0].expr() * expr_256.clone() + c.cells[0].expr();
+                // a + b (first carry_in is always 0)
+                let rhs = a.cells[0].expr() + b.cells[0].expr();
+                // equality check
+                constraints.push(lhs - rhs);
+
+                for idx in 1..32 {
+                    // 256 * carry_out + c
+                    let lhs = carry[idx].expr() * expr_256.clone()
+                        + c.cells[idx].expr();
+                    // a + b + carry_in
+                    let rhs = a.cells[idx].expr()
+                        + b.cells[idx].expr()
+                        + carry[idx - 1].expr();
+                    // equality check
+                    constraints.push(lhs - rhs)
+                }
+
+                constraints
+            };
+
+            #[allow(clippy::suspicious_operation_groupings)]
+            let bus_mapping_lookups = vec![
+                Lookup::BusMappingLookup(BusMappingLookup::Stack {
+                    index_offset: 0,
+                    value: a.expr(),
+                    is_write: false,
+                }),
+                Lookup::BusMappingLookup(BusMappingLookup::Stack {
+                    index_offset: 1,
+                    value: swap.expr() * c.expr() + no_swap.clone() * b.expr(),
+                    is_write: false,
+                }),
+                Lookup::BusMappingLookup(BusMappingLookup::Stack {
+                    index_offset: 1,
+                    value: swap.expr() * b.expr() + no_swap * c.expr(),
+                    is_write: true,
+                }),
+            ];
+
+            Constraint {
+                name: "AddGadget success",
+                selector: selector.expr(),
+                polys: [
+                    state_transition_constraints,
+                    swap_constraints,
+                    add_constraints,
+                ]
+                .concat(),
+                lookups: bus_mapping_lookups,
+            }
+        };
+
+        let stack_underflow = {
+            let (zero, minus_one) = (
+                Expression::Constant(F::from_u64(1024)),
```

Comment:
stack pointer should range from 0...1023,  check 1024 is redundant ?

Reply:
I think you got a point there! @han0110 It seems correct to me. We should have 1023 as max.

Reply:
In evm circuit, we have `stack_pointer` initialized at 1024, which represents it's empty. For handling `ADD`, we need to pop 2 values from stack, which means we have 2 kinds of stack underflow condition: when stack is empty or stack only contains 1 item. We need to check both conditions, right?

Reply:
seems right that stack_pointer initialized at 1024, which represents point to  empty state, pointing to 1023 means one item in the stack,  do we have code to ensure initialized  stack_pointer?

Reply:
Well I thought that the stack has 1024 elements. Therefore, we check range [0-1023] which is 1024 elements inside since 0 counts.

We also work on the same way in the Stack repr of the crate. The stack pointer points at the beginning of the execution to 1023 and goes down to 0 (being 0 the last slot to allocate in the stack).

Reply:
`CallInit` I am currently working on initializes `OpExecutionState`.